data-serving/scripts/convert-data/geocode_util.py

```diff
@@ -0,0 +1,200 @@
+        # etc. Ex. "Paris", "China", "FL"
+        return lookup_single_part_location(geocoder, normalized_tokens[0])
+    if len(normalized_tokens) == 2:
+        # A two-part location, which may be (county, country), (city, county),
+        # or (city, country) pairs. Ex. "Paris, France", "Hubei, China",
+        # "Boston, MA"
+        return lookup_two_part_location(
+            geocoder, normalized_tokens[0],
+            normalized_tokens[1])
+    if len(normalized_tokens) == 3:
+        # A three-part location, presumed to including city, county, and
+        # country.
+        return geocoder.geocode(
+            normalized_tokens[0],  # city
+            normalized_tokens[1],  # county
+            normalized_tokens[2])  # country
+
+    raise ValueError('Too many tokens in location')
+
+
+def lookup_two_part_location(
+        geocoder: Any, higher_res_location_token: str,
+        lower_res_location_token: str) -> Any:
+    '''
+    Attempts to match the two location tokens against known locations,
+    trying two-token pairs from broadest to most specific resolution. First we
+    check if it's a province and county pair; then city and country; then city
+    and province.
+
+    Parameters:
+        higher_res_location_token: The token representing the more specific part
+          of the location, e.g. city vs. country.
+        lower_res_location_token: The token representing the broader part of the
+          location, e.g. county vs. city.
+
+    Returns:
+      None: When no geocoding match is found.
+      [Dict[str, Any]]: Location data of the form:
+        {
+          lat: float
+          lng: float
+          geo_resolution: str
+          country_new: str
+          admin_id: int
+          location: str
+          admin3: str
+          admin2: str
+          admin1: str
+        }
+    '''
+    province_and_country = lookup(
+        geocoder, lambda location: location.admin1.lower() ==
+        higher_res_location_token and location.country_new.lower() ==
+        lower_res_location_token and location.geo_resolution == 'admin1')
+
+    if province_and_country:
+        return province_and_country
+
+    city_and_country = lookup(
+        geocoder, lambda location: location.admin2.lower() ==
+        higher_res_location_token and location.country_new.lower() ==
+        lower_res_location_token and location.geo_resolution == 'admin2')
+
+    if city_and_country:
+        return city_and_country
+
+    city_and_province = lookup(
+        geocoder, lambda location: location.admin2.lower() ==
+        higher_res_location_token and location.admin1.lower() ==
+        lower_res_location_token and location.geo_resolution == 'admin2')
+
+    if city_and_province:
+        return city_and_province
+
+    return None
+
+
+def lookup_single_part_location(geocoder: Any, value: str) -> Any:
+    '''
+    Attempts to match the location token against known locations, beginning from
```

Comment:
Would the "Paris" confusion be a result of granularity direction? IOW, aren't both Paris cities? Are we representing likelihood some other way?

Reply:
Related: https://daringfireball.net/2020/05/what_time_is_it_in_london. Paris, TX and Paris, France are indeed both cities, and I don't know that it's a good idea to assume any particular Paris if it's not clear from context. There are 23 Parises in the US and a couple in Canada.

My reading of this code is that it does a right thing (I'm not feeling qualified to say _the_ right thing, but definitely _a_ right thing): if we get here with the value `Paris`, then we don't find a country or province called Paris, then we find a bajillion towns called Paris and raise `ValueError`.

However, if we get here because the single location token is `Georgia`, we'll choose the country as a coarser grain than the state, and I don't know if that's correct.

Reply:
OK I did pick a terrible, terrible example -- in reality, Paris, TX doesn't actually raise an issue here because it's not in this location db -- I'll clean up the comment

*However*: the bigger question you raised is good, and I've been back and forth on it. @iamleeg your example about Georgia is spot on. I was trying to save cycles by greedily matching but it's probably not worth the loss in precision, so I'll attempt all matches for the # of tokens, and if there's more than one result, I'll throw an error about disambiguation.

The problem with this approach is that there isn't really a way to "clarify" Georgia the country -- when I send the logs to the curators to fix ambiguous locations, what would they put in there to indicate that it's NOT the state? We'll have to figure out how to fix this and I'm open to suggestions.

Reply:
I saw some "convenient" abbreviations in `constants.py`, can the ISO 3166-2 country code for Georgia (`ge`) be added? Then `ge` matches Georgia the country, `ga` matches Georgia, US, and hopefully that's the only example where we need to make people do that.

Sadly `ga` is the ISO code for Gabon, but we can ask people to write "Gabon" when they mean that. Gah!

Reply:
I think this will take a bit more work because of the disambiguation errors now -- 'ge' will return 'georgia', but then we will raise an error because we don't know if 'georgia' is the country or the state. So I think I may need to rework the abbreviations, perhaps to map to the correct result ID instead of just an abbreviation (brittle, maybe, but more precise). I'll send a separate PR for this today since this changes the flow in `geocode_util`.